packages/multichain/src/caip25Permission.ts

```diff
@@ -70,6 +72,46 @@ type Caip25EndowmentCaveatSpecificationBuilderOptions = {
   listAccounts: () => { address: Hex }[];
 };
 
+/**
+ * Calculates the difference between two provided CAIP-25 permission caveat values, but only considering a single scope property at a time.
+ *
+ * @param originalValue - The existing CAIP-25 permission caveat value.
+ * @param mergedValue - The result from merging existing and incoming CAIP-25 permission caveat values.
+ * @param scopeToDiff - The required or optional scopes from the [CAIP-25](https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-25.md) request.
+ * @returns The difference between original and merged CAIP-25 permission caveat values.
+ */
+export function diffScopesForCaip25CaveatValue(
+  originalValue: Caip25CaveatValue,
+  mergedValue: Caip25CaveatValue,
+  scopeToDiff: 'optionalScopes' | 'requiredScopes',
+): Caip25CaveatValue {
+  const diff = cloneDeep(originalValue);
+
+  for (const [scopeString, mergedScopeObject] of Object.entries(
+    mergedValue[scopeToDiff],
+  )) {
+    const internalScopeString = scopeString as InternalScopeString;
```

Comment:
Nit: I'm always on the lookout for ways to avoid casting like this, because it can be really easy to introduce type errors this way. But the type of keys always gets collapsed to `string`, so I can see why it's necessary here.

However, here is an idea to make this _slightly_ safer, so that the key type stays correct even if the type of the collection changes:

```suggestion
  const mergedScopeToDiff = mergedValue[scopeToDiff];
  for (const [scopeString, mergedScopeObject] of Object.entries(
    mergedScopeToDiff,
  )) {
    const internalScopeString = scopeString as keyof typeof mergedScopeToDiff;
```

Reply:
Very good solution! Thank you 🙏🏾 
addressed in https://github.com/MetaMask/core/pull/5283/commits/4947852e2a1be8c761b78074913698cc144eec5a